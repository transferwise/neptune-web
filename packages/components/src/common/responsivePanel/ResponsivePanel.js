import React, { forwardRef } from 'react';
import Types from 'prop-types';

import { Position } from '..';
import BottomSheet from '../bottomSheet';
import Panel from '../panel';
import SizeSwapper from '../../sizeSwapper';

const ResponsivePanel = forwardRef(({ arrow, children, open, position, triggerRef }, ref) => {
  const windowRef = typeof window !== 'undefined' && window;

  const items = [
    {
      items: [
        <BottomSheet open={open} ref={ref} key="bottomSheet">
          {children}
        </BottomSheet>,
      ],
    },
    {
      items: [
        <Panel
          arrow={arrow}
          open={open}
          position={position}
          ref={ref}
          triggerRef={triggerRef}
          key="panel"
        >
          {children}
        </Panel>,
      ],
      breakpoint: SizeSwapper.Breakpoint.SMALL,
    },
  ];
  return <SizeSwapper items={items} ref={windowRef} />;
});

ResponsivePanel.Position = {
  BOTTOM: Position.BOTTOM,
  LEFT: Position.LEFT,
  RIGHT: Position.RIGHT,
  TOP: Position.TOP,
};

ResponsivePanel.defaultProps = {
  position: ResponsivePanel.Position.TOP,
  arrow: false,
};

ResponsivePanel.propTypes = {
  children: Types.node.isRequired,
  position: Types.oneOf([
    ResponsivePanel.Position.TOP,
    ResponsivePanel.Position.RIGHT,
    ResponsivePanel.Position.LEFT,
    ResponsivePanel.Position.BOTTOM,
  ]),
  arrow: Types.bool,
  open: Types.bool.isRequired,
  triggerRef: Types.shape({ current: Types.shape({}) }).isRequired,
};

export default ResponsivePanel;
