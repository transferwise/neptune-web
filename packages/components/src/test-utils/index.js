import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
