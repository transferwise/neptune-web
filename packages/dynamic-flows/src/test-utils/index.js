import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
