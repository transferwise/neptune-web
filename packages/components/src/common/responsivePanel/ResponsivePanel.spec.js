import React from 'react';
import { fireEvent, render, waitFor } from '../../test-utils';

import ResponsivePanel from './ResponsivePanel';
import { Breakpoint } from '..';

jest.mock('../bottomSheet', () => {
  // eslint-disable-next-line
  const { forwardRef } = require('react');
  // eslint-disable-next-line react/prop-types
  return forwardRef(({ open, children }, ref) =>
    open ? (
      <div ref={ref} className="np-bottom-sheet">
        {children}
      </div>
    ) : null,
  );
});

jest.mock('../panel', () => {
  // eslint-disable-next-line
  const { forwardRef } = require('react');
  // eslint-disable-next-line react/prop-types
  return forwardRef(({ open, children }, ref) =>
    open ? (
      <div ref={ref} className="np-panel np-panel--open">
        {children}
      </div>
    ) : null,
  );
});

describe('ResponsivePanel', () => {
  const props = {
    arrow: true,
    children: 'content',
    open: true,
    position: ResponsivePanel.Position.TOP,
    title: 'title',
    anchorRef: jest.fn(),
    onClose: jest.fn(),
  };
  afterEach(() => {
    jest.clearAllMocks();
  });

  let container;

  describe('on desktop', () => {
    it('renders Panel', async () => {
      await waitFor(() => {
        ({ container } = render(
          <ResponsivePanel {...props} intialOpen>
            children
          </ResponsivePanel>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('calls onClose with correct value', async () => {
      await waitFor(() => {
        render(<ResponsivePanel {...props}>children</ResponsivePanel>);
      });

      expect(props.onClose).not.toHaveBeenCalled();

      fireEvent.click(getPanel());

      expect(props.onClose).not.toHaveBeenCalled();

      fireEvent.click(document);

      expect(props.onClose).toHaveBeenCalled();
    });
  });

  describe('on mobile', () => {
    beforeEach(() => {
      global.innerWidth = Breakpoint.SMALL - 1;
    });

    it('renders bottomSheet', async () => {
      await waitFor(() => {
        ({ container } = render(
          <ResponsivePanel {...props} open>
            children
          </ResponsivePanel>,
        ));
      });

      expect(container).toMatchSnapshot();
    });

    it('calls onClose with correct value', async () => {
      await waitFor(() => {
        render(<ResponsivePanel {...props}>children</ResponsivePanel>);
      });

      expect(props.onClose).not.toHaveBeenCalled();

      fireEvent.click(getBottomSheet());

      expect(props.onClose).not.toHaveBeenCalledWith();

      fireEvent.click(document);

      expect(props.onClose).toHaveBeenCalled();
    });
  });

  const getBottomSheet = () => document.querySelector('.np-bottom-sheet');
  const getPanel = () => document.querySelector('.np-panel');
});
