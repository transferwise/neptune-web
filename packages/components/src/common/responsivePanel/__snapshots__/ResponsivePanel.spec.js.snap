// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsivePanel on desktop renders Panel 1`] = `
<div>
  <div
    class="np-size-swapper d-flex"
    style="visibility: visible;"
  >
    <div
      class="np-panel np-panel--open"
    >
      children
    </div>
  </div>
</div>
`;

exports[`ResponsivePanel on mobile renders bottomSheet 1`] = `
<div>
  <div
    class="np-size-swapper d-flex"
    style="visibility: visible;"
  >
    <div
      class="np-bottom-sheet"
    >
      children
    </div>
  </div>
</div>
`;
